torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
   return at::linalg_solve(Uh, at::matmul(U_contrib, Uh) + D_contrib * Uh);
 }
 
-// http://eprints.maths.ox.ac.uk/1079/1/NA-08-01.pdf
-Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
-                    bool eigenvectors, bool upper, const Tensor& lambda, const Tensor& v) {
-  // This gradient is symmetric, and not triangular.
-  // symeig operates only on symmetric inputs, which is a subspace of
-  // R^{n x n}, and hence the derivative is not well-defined for off-diagonal
-  // elements. We resolve this by taking the gradient of the functionally independent
-  // elements of the matrix (i.e., the lower triangular portion of the input) and then
-  // reflect it on the upper triangular portion, thereby symmetrizing the gradient of
-  // the symeig operation. The motivation behind this choice is that symmetric gradient
-  // leads to stable gradient updates, and retains symmetry of the updated matrix if it
-  // were updated by a gradient based algorithm.
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+Tensor eigh_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
+                     bool eigenvectors, const Tensor& lambda, const Tensor& v) {
+  // This function is used for both torch.symeig and torch.linalg.eigh.
+  // eigh (and torch.symeig) operates only on symmetric (resp. Hermitian) inputs.
+  // This is a _linear_ subspace of R^{n x n} (resp. C^{n x n}),
+  // since A, B \in Sym(n) => A + B \in Sym(n)
+  // This means that, the tangent space to Sym(n) is Sym(n),
+  // so the gradient is symmetric (resp. Hermitian).
+  // For this reason, we disregard the `upper` flag in torch.symeig and the `UPLO` flag
+  // in torch.linalg.eigh, and we simply work with symmetric matrices.
+
+  // This check just can be triggered in the backwards of torch.symeig
   TORCH_CHECK(eigenvectors,
-           "symeig_backward: Setting eigenvectors to false in torch.symeig doesn't compute eigenvectors ",
-           "and hence we cannot compute backward. Please use torch.symeig(eigenvectors=True)");
+           "Setting eigenvectors=False in torch.symeig doesn't compute the eigenvectors, ",
+           "and hence we cannot compute the backward. Please use torch.symeig(eigenvectors=True)");
 
   auto glambda = grads[0];
   auto gv = grads[1];
 
   auto vh = v.conj().transpose(-2, -1);
 
+  if (glambda.defined()) {
+    // glambda is real even when the matrix is complex
+    glambda = glambda.to(self.dtype());
+  }
```

Comment:
It could be that there is no need to convert `glambda` to complex at all.
Elementwise products of complex-real tensors work and also `copy` seems to work as well:
```python
In [2]: import torch

In [3]: x = torch.rand(3, 3, dtype=torch.cdouble)

In [4]: y = torch.rand(3, 3, dtype=torch.double)

In [5]: x
Out[5]: 
tensor([[0.5320+0.1139j, 0.9173+0.0720j, 0.8534+0.6649j],
        [0.0010+0.0615j, 0.2283+0.1710j, 0.8142+0.7606j],
        [0.3496+0.3054j, 0.3624+0.1556j, 0.0159+0.5936j]],
       dtype=torch.complex128)

In [6]: x.copy_(y)
Out[6]: 
tensor([[0.9531+0.j, 0.6419+0.j, 0.2485+0.j],
        [0.2854+0.j, 0.4344+0.j, 0.2342+0.j],
        [0.0689+0.j, 0.7904+0.j, 0.3975+0.j]], dtype=torch.complex128)

In [7]: x
Out[7]: 
tensor([[0.9531+0.j, 0.6419+0.j, 0.2485+0.j],
        [0.2854+0.j, 0.4344+0.j, 0.2342+0.j],
        [0.0689+0.j, 0.7904+0.j, 0.3975+0.j]], dtype=torch.complex128)

```

Reply:
Good point!